pkg/service/service.go

```diff
@@ -1,46 +1,348 @@
+						},
+						func(qs queryState) queryState {
+							qs.qr.Indexes.Set(result.ContextID, index)
+							return qs
+						})
+
+					// add location queries for all shards containing the original CID we're seeing an index for
+					shards := index.Shards().Iterator()
+					for shard, index := range shards {
+						if index.Has(*j.indexForMh) {
+							if err := spawn(job{shard, nil, nil, equalsOrLocationJobType}); err != nil {
+								return err
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// Query returns back relevant content claims for the given query using the following steps
+// 1. Query the IPNIIndex for all matching records
+// 2. For any index records, query the IPNIIndex for any location claims for that index cid
+// 3. For any index claims, query the IPNIIndex for location claims for the index cid
+// 4. Query the BlobIndexLookup to get the full ShardedDagIndex for any index claims
+// 5. Query IPNIIndex for any location claims for any shards that contain the multihash based on the ShardedDagIndex
+// 6. Read the requisite claims from the ClaimLookup
+// 7. Return all discovered claims and sharded dag indexes
+func (is *IndexingService) Query(ctx context.Context, q Query) (queryresult.QueryResult, error) {
+	initialJobs := make([]job, 0, len(q.Hashes))
+	for _, mh := range q.Hashes {
+		initialJobs = append(initialJobs, job{mh, nil, nil, standardJobType})
+	}
+	qs, err := is.jobWalker(ctx, initialJobs, queryState{
+		q: &q,
+		qr: &queryResult{
+			Claims:  make(map[cid.Cid]delegation.Delegation),
+			Indexes: bytemap.NewByteMap[types.EncodedContextID, blobindex.ShardedDagIndexView](-1),
+		},
+		visits: map[jobKey]struct{}{},
+	}, is.jobHandler)
+	if err != nil {
+		return nil, err
+	}
+	return queryresult.Build(qs.qr.Claims, qs.qr.Indexes)
+}
+
+func (is *IndexingService) urlForResource(provider peer.AddrInfo, resourceType string, resourceID string) (*url.URL, error) {
+	for _, addr := range provider.Addrs {
+		// first, attempt to convert the addr to a url scheme
+		url, err := maurl.ToURL(addr)
+		// if it can't be converted, skip
+		if err != nil {
+			continue
+		}
+		// must be an http url
+		if !(url.Scheme == "http" || url.Scheme == "https") {
+			continue
+		}
+		// we must have a place to place the resourceId in the path
+		if !strings.Contains(url.Path, resourceType) {
+			continue
+		}
+		// ok we have a matching URL, return with all resource type components replaced with the id
+		url.Path = strings.ReplaceAll(url.Path, resourceType, resourceID)
+		return url, nil
+	}
+	return nil, errors.New("no claim endpoint found")
+}
+
+func (is *IndexingService) fetchClaimURL(provider peer.AddrInfo, claimCid cid.Cid) (*url.URL, error) {
+	return is.urlForResource(provider, "{claim}", claimCid.String())
+}
+
+func (is *IndexingService) fetchRetrievalURL(provider peer.AddrInfo, shard cid.Cid) (*url.URL, error) {
+	return is.urlForResource(provider, "{shard}", shard.String())
+}
```

Comment:
@alanshaw relevant for your IPNI work. See https://github.com/ipni/go-libipni/issues/230